### PR TITLE
RHICOMPL-3021 use same build logic for pr_check and build_deploy

### PR DESCRIPTION
**Pull request by Victoremepunto**
- Refactor build-deploy-common deployment library
- Deprecate backwards compatibility tags (qa, latest tags no longer used)
- Update naming format of temporary image tags to match bonfire's build script
- Enforce usage of `docker` where available, fallback to `podman` when it's not.
- Detect and fail if it can't build the image

## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [x] General Coding Practices

### CODE REVIEW

**Comment by InsightsDroid** [comment]
Test results are available in [Ibutsu](https://url.corp.redhat.com/ibutsu-runs). The test run IDs are:
484f2583-2919-457a-8ea4-0de772464589
56737b20-d9a0-4068-ac4a-93cc539580ce

File: deploy/build-deploy-common.sh [hunk 1]
```diff
@@ -1,35 +1,55 @@
...
+additional_tags() {
+  [ -n "$ADDITIONAL_TAGS" ]
 }
 
 is_ci_runner() {
     [[ "$CI" == true ]]
 }
 
+is_pr_or_mr_build() {
+    [ -n "$ghprbPullId" ] || [ -n "$gitlabMergeRequestIid" ]
```
Comment (romanblanco) [inline deploy/build-deploy-common.sh:30]:
```suggestion
    [ -n "$ghprbPullId" ] || [ -n "$gitlabMergeRequestIid" ]
```
```suggestion
    [ -n "$ghprbPullId" ] || [ -n "$gitlabMergeRequestId" ]
```

File: deploy/build-deploy-common.sh [hunk 2]
```diff
@@ -1,35 +1,55 @@
...
+}
+
+get_pr_build_id() {
+
+    local BUILD_ID
+
+    if [ -n "$ghprbPullId" ]; then
+        BUILD_ID="$ghprbPullId"
+    elif [ -n "$gitlabMergeRequestIid" ]; then
```
Comment (romanblanco) [inline deploy/build-deploy-common.sh:39]:
```suggestion
    elif [ -n "$gitlabMergeRequestId" ]; then
```

...and others

Comment (romanblanco) [inline deploy/build-deploy-common.sh:39]:
```suggestion
    elif [ -n "$gitlabMergeRequestId" ]; then
```

...and others

File: deploy/build-deploy-common.sh [hunk 3]
```diff
@@ -157,14 +179,27 @@ _get_build_args() {
...
-        container_engine_cmd build --pull -f "$DOCKERFILE" $BUILD_ARGS_CMD -t "${IMAGE_NAME}:${IMAGE_TAG}" .
-    else
-        container_engine_cmd build --pull -f "$DOCKERFILE" -t "${IMAGE_NAME}:${IMAGE_TAG}" .
     fi
 
+    container_engine_cmd build --pull -f "$DOCKERFILE" $BUILD_ARGS_CMD $LABEL_PARAMETER -t "${IMAGE_NAME}:${IMAGE_TAG}" .
+
+    if [ $? != 0 ]; then
+        echo "Error building image"
+	return 1
```
Comment (romanblanco) [inline deploy/build-deploy-common.sh:197]: nitpick: indentation missing

Comment (romanblanco) [inline deploy/build-deploy-common.sh:197]: nitpick: indentation missing

**Comment by vkrizan** [comment]
> Deprecate backwards compatibility tags (qa, latest tags no longer used)

Can the `latest` tag still be used and set on latest build from main/master?

> Update naming format of temporary image tags to match bonfire's build script

Is it possible to minimize custom parts and use bonfire's build script to the fullest?

> Enforce usage of docker where available, fallback to podman when it's not.

I'd prefer if it would be the other way around - use `podman` exclusively. AFAIK is is available since RHEL8. Jenkins node would need to be changed.

@Victoremepunto thx

**Comment by skateman** [comment]
What's the status on this? :thinking: :watch:

**Comment by Victoremepunto** [comment]
> What's the status on this? thinking watch

Same as #1284  , need to address @vkrizan  concerns. inline: 

> Can the latest tag still be used

Can do whatever we want - but - if we don't need it I don't see the point. I'll add it as an "additional tag" 

> Is it possible to minimize custom parts 

The usage of this "build-deploy-common" was to exactly reduce duplicity, and it was there way before the bonfire tooling grew up as it is now (it didn't have all the features it has today, is what I mean). The build-deploy-common handles also different container engines with less duplicity, also the `build` logic for images supports multi stage builds, which is not supported to my knowledge by bonfire build scripts. 

This PR updates `build-deploy-common` to match bonfire's tag format and fixes a couple of quirks. We can:
 - accept this changes and move on using build-deploy-common
 - drop build-deploy-common and just use bonfire scripts (it would imply changes in compliance-ssg as well, and I think floorist too).

We should decide on what to do. 

I'm probably biased, because I wrote build-deploy-common , but. I personally find "nice" that by using build-deploy-common, the "only" thing you need to specify is a couple of variables and you're done. Just compare:
- [compliance-backend build-deploy](https://github.com/RedHatInsights/compliance-backend/blob/master/build_deploy.sh)
- [floorist build_deploy](https://github.com/RedHatInsights/floorist/blob/master/build_deploy.sh)
and compliance-ssg's (not gonna put the url here)

with 
- [inventory's build-deploy](https://github.com/RedHatInsights/insights-host-inventory/blob/master/build_deploy.sh)
- [sources build-deploy](https://github.com/RedHatInsights/sources-api-go/blob/main/build_deploy.sh)

just to name a few. I kind of think that I like this idea more, but LMK if you don't agree.

> Enforce usage of docker where available, fallback to podman when it's not.

This was done because [this](https://github.com/RedHatInsights/compliance-backend/pull/938/commits/45cf2c30254c326ed9e237858d420d1b0afa2e6d)  introduced that in the build script to force docker on purpose for some bug , I believe. I'm happy to fallback use `podman` by default. Actually, this is the exact reason why the build-deploy-common couldn't be bumped on Compliance-SSG repo , there's an MR there waiting for this changes to land so we can have the version updated.

**Comment by vkrizan** [comment]
> > Can the latest tag still be used
> 
> Can do whatever we want - but - if we don't need it I don't see the point. I'll add it as an "additional tag"
> 

Thanks. It's a best practice to tag `latest`. AFAIK, some defect tracking sw are looking mainly at that tag.

> > Is it possible to minimize custom parts
> 
> The usage of this "build-deploy-common" was to exactly reduce duplicity, and it was there way before the bonfire tooling grew up as it is now (it didn't have all the features it has today, is what I mean). The build-deploy-common handles also different container engines with less duplicity, also the `build` logic for images supports multi stage builds, which is not supported to my knowledge by bonfire build scripts.
> 
> This PR updates `build-deploy-common` to match bonfire's tag format and fixes a couple of quirks. We can:
> 
>     * accept this changes and move on using build-deploy-common
> 
>     * drop build-deploy-common and just use bonfire scripts (it would imply changes in compliance-ssg as well, and I think floorist too).
> 
> 
> We should decide on what to do.
> 
> I'm probably biased, because I wrote build-deploy-common , but. I personally find "nice" that by using build-deploy-common, the "only" thing you need to specify is a couple of variables and you're done. 
> I kind of think that I like this idea more, but LMK if you don't agree.

TBH I'd wanna see use of the common bonfire CI scripts as much as possible. It does not have to be part of this PR, but I think we should strive towards that. Additional logic of our common script could be contributed to bonfire's repo. Everything else/specific to our case might be kept in ours.

> 
> > Enforce usage of docker where available, fallback to podman when it's not.
> 
> This was done because [this](https://github.com/RedHatInsights/compliance-backend/pull/938/commits/45cf2c30254c326ed9e237858d420d1b0afa2e6d) introduced that in the build script to force docker on purpose for some bug , I believe. I'm happy to fallback use `podman` by default. Actually, this is the exact reason why the build-deploy-common couldn't be bumped on Compliance-SSG repo , there's an MR there waiting for this changes to land so we can have the version updated.

We can switch the Jenkins build node to rhel8 which (only) has podman.

**Comment by vkrizan** [comment]
This PR enlists 5 issues being resolved, but there is just one commit. It is always best to separate each concern, for the sake of review or possible revert.

File: deploy/build-deploy-common.sh [hunk 4]
```diff
@@ -1,35 +1,55 @@
...
+}
+
+get_pr_build_id() {
+
+    local BUILD_ID
+
+    if [ -n "$ghprbPullId" ]; then
+        BUILD_ID="$ghprbPullId"
+    elif [ -n "$gitlabMergeRequestId" ]; then
+	BUILD_ID="$gitlabMergeRequestId"
```
Comment (vkrizan) [inline deploy/build-deploy-common.sh:40]: Indentation.

File: deploy/build-deploy-common.sh [hunk 5]
```diff
@@ -157,14 +179,27 @@ _get_build_args() {
...
 
+    container_engine_cmd build --pull -f "$DOCKERFILE" $BUILD_ARGS_CMD $LABEL_PARAMETER -t "${IMAGE_NAME}:${IMAGE_TAG}" .
+
+    if [ $? != 0 ]; then
+        echo "Error building image"
+	return 1
+    fi
+}
+
+get_expiry_label_parameter() {
```
Comment (vkrizan) [inline deploy/build-deploy-common.sh:201]: This might be a constant (for the duration of the script).

File: deploy/build-deploy-common.sh [hunk 6]
```diff
@@ -208,23 +244,31 @@ build_deploy_init() {
...
 }
 
+set_image_tag() {
+
+    local BUILD_ID
+
+    BUILD_ID=$(get_pr_build_id)
+
+    if [ -n "$BUILD_ID" ]; then
+    IMAGE_TAG="pr-${BUILD_ID}-$(get_7_chars_commit_hash)"
```
Comment (vkrizan) [inline deploy/build-deploy-common.sh:254]: Indentation.

**Comment by Victoremepunto** [comment]
> This PR enlists 5 issues being resolved, but there is just one commit. It is always best to separate each concern, for the sake of review or possible revert.

TBH I no longer remember where we stand as of today for the commit and PR format, I personally prefer to sqash everything into 1 commit detailing what the commit does - the "[CONTRIBUTING](https://github.com/RedHatInsights/compliance-backend#contributing)" section is very vague on that respect, maybe we could update it to reflect what is the expectation for future PRs ?

**Comment by Victoremepunto** [comment]
> > > Can the latest tag still be used
> > 
> > 
> > Can do whatever we want - but - if we don't need it I don't see the point. I'll add it as an "additional tag"
> 
> Thanks. It's a best practice to tag `latest`. AFAIK, some defect tracking sw are looking mainly at that tag.
> 

I've restored the `latest` additional tag so it's created only when it's not a PR or an MR event 

> > > Is it possible to minimize custom parts
> > 
> > 
> > The usage of this "build-deploy-common" was to exactly reduce duplicity, and it was there way before the bonfire tooling grew up as it is now (it didn't have all the features it has today, is what I mean). The build-deploy-common handles also different container engines with less duplicity, also the `build` logic for images supports multi stage builds, which is not supported to my knowledge by bonfire build scripts.
> > This PR updates `build-deploy-common` to match bonfire's tag format and fixes a couple of quirks. We can:
> > ```
> > * accept this changes and move on using build-deploy-common
> > 
> > * drop build-deploy-common and just use bonfire scripts (it would imply changes in compliance-ssg as well, and I think floorist too).
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > We should decide on what to do.
> > I'm probably biased, because I wrote build-deploy-common , but. I personally find "nice" that by using build-deploy-common, the "only" thing you need to specify is a couple of variables and you're done.
> > I kind of think that I like this idea more, but LMK if you don't agree.
> 
> TBH I'd wanna see use of the common bonfire CI scripts as much as possible. It does not have to be part of this PR, but I think we should strive towards that. Additional logic of our common script could be contributed to bonfire's repo. Everything else/specific to our case might be kept in ours.
> 
I don't know how we can keep a balance like that using bonfire's scripts and keeping our specific use cases covered - I'm afraid we'll be duplicating more code if this functionality is not accepted on bonfire's scripts (the stage builds for example, or the compatibility with local builds). 

If this can be done outside of this PR, I can "refrain" from adding more functionality to the build-deploy-common library and migrate to bonfire-scripts only just like anyone else if that's what the team wants to do.

> > > Enforce usage of docker where available, fallback to podman when it's not.
> > 
> > 
> > This was done because [this](https://github.com/RedHatInsights/compliance-backend/pull/938/commits/45cf2c30254c326ed9e237858d420d1b0afa2e6d) introduced that in the build script to force docker on purpose for some bug , I believe. I'm happy to fallback use `podman` by default. Actually, this is the exact reason why the build-deploy-common couldn't be bumped on Compliance-SSG repo , there's an MR there waiting for this changes to land so we can have the version updated.
> 
> We can switch the Jenkins build node to rhel8 which (only) has podman.

Well, I've just hit one of the issues we were actually hitting by using `podman`. 

```
10:15:00 Error: error creating build container: Error committing the finished image: error adding layer with blob "sha256:6c1f16d57732ffd28227953136d0395a1a2a7a269c8dbb2759820294fc0072c8": Error processing tar file(exit status 1): there might not be enough IDs available in the namespace (requested 0:12 for /var/spool/mail): lchown /var/spool/mail: invalid argument
```

The quick work-around for this is to use `docker`. We've pinged AppSRE in the past about this and they fixed it, though not permanently, it seems. defaulting to `docker` , sadly, seems for reliable for the time being. thoughts ?

**Comment by Victoremepunto** [comment]
> Well, I've just hit one of the issues we were actually hitting by using `podman`.
> 
> ```
> 10:15:00 Error: error creating build container: Error committing the finished image: error adding layer with blob "sha256:6c1f16d57732ffd28227953136d0395a1a2a7a269c8dbb2759820294fc0072c8": Error processing tar file(exit status 1): there might not be enough IDs available in the namespace (requested 0:12 for /var/spool/mail): lchown /var/spool/mail: invalid argument
> ```
> 
> The quick work-around for this is to use `docker`. We've pinged AppSRE in the past about this and they fixed it, though not permanently, it seems. defaulting to `docker` , sadly, seems for reliable for the time being. thoughts ?

I've hit another one :man_facepalming: :see_no_evil:   , check build 2153

**Comment by vkrizan** [comment]
> I've restored the latest additional tag so it's created only when it's not a PR or an MR event

:+1: ty

> If this can be done outside of this PR, I can "refrain" from adding more functionality to the build-deploy-common library and migrate to bonfire-scripts only just like anyone else if that's what the team wants to do.

:+1: 

> TBH I no longer remember where we stand as of today for the commit and PR format, I personally prefer to sqash everything into 1 commit detailing what the commit does - the "[CONTRIBUTING](https://github.com/RedHatInsights/compliance-backend#contributing)" section is very vague on that respect, maybe we could update it to reflect what is the expectation for future PRs ?

General Git Best practices. We have an internal page from AppSRE. I think we can include the portion of it to CONTRIBUTING file.

> I've hit another one man_facepalming see_no_evil , check build 2153

Those errors need to be figured out. I suggest to start by changing the Jenkins node to `rhel8`. (check appinterface config for build)

**Comment by Victoremepunto** [comment]
> > TBH I no longer remember where we stand as of today for the commit and PR format, I personally prefer to sqash everything into 1 commit detailing what the commit does - the "[CONTRIBUTING](https://github.com/RedHatInsights/compliance-backend#contributing)" section is very vague on that respect, maybe we could update it to reflect what is the expectation for future PRs ?
> 
> General Git Best practices. We have an internal page from AppSRE. I think we can include the portion of it to CONTRIBUTING file.

ACK. Good item for our next retrospective meeting. Thanks.

> > I've hit another one man_facepalming see_no_evil , check build 2153
> 
> Those errors need to be figured out. I suggest to start by changing the Jenkins node to `rhel8`. (check appinterface config for build)

ACK - These errors have been reported to ops. team, and as discussed we're moving to use RHEL8 nodes for our builds

**Comment by Victoremepunto** [comment]
> > > TBH I no longer remember where we stand as of today for the commit and PR format, I personally prefer to sqash everything into 1 commit detailing what the commit does - the "[CONTRIBUTING](https://github.com/RedHatInsights/compliance-backend#contributing)" section is very vague on that respect, maybe we could update it to reflect what is the expectation for future PRs ?
> > 
> > 
> > General Git Best practices. We have an internal page from AppSRE. I think we can include the portion of it to CONTRIBUTING file.
> 
> ACK. Good item for our next retrospective meeting. Thanks.
> 
> > > I've hit another one man_facepalming see_no_evil , check build 2153
> > 
> > 
> > Those errors need to be figured out. I suggest to start by changing the Jenkins node to `rhel8`. (check appinterface config for build)
> 
> ACK - These errors have been reported to ops. team, and as discussed we're moving to use RHEL8 nodes for our builds

@skateman  @vkrizan  okay, the RHEL8 node issues with our unit_test.sh script are gone now with #1367 
anything pending here now ?

**Pull request merged**